### PR TITLE
refactor: update trivial regex capture group find and replace challenge

### LINKED ISSUE #17645: regex search and replace challenge is trivial
#### Describe your problem and - if possible - how to reproduce it
Instead of testing anything useful that is mentioned in the description, it requires a plaintext replacement, not even needing a capture group.

I suggest it is changed to a reordering exercise, since it is already mentioned in the example.

#### Add a Link to the page with the problem
https://learn.freecodecamp.org/javascript-algorithms-and-data-structures/regular-expressions/use-capture-groups-to-search-and-replace/

### PR DESCRIPTION
**Pull request by PartyLich**
Implements the fixes provided by @alanpaulprice in pull #36925 with
the changes requested by maintainers. Previous challenge could be
completed without any capture groups. New challenge rectifies this
behavior.

resolves #17645

- [x] I have read [freeCodeCamp's contribution guidelines](https://github.com/freeCodeCamp/freeCodeCamp/blob/master/CONTRIBUTING.md).
- [x] My pull request has a descriptive title (not a vague title like `Update index.md`)
- [x] My pull request targets the `master` branch of freeCodeCamp.
- [x] None of my changes are plagiarized from another source without proper attribution.
   Note: as the commit text says, this is the pull referenced above with the requested corrections. Not my work, basically.
- [x] All the files I changed are in the same world language (for example: only English changes, or only Chinese changes, etc.)
- [x] My changes do not use shortened URLs or affiliate links.

Closes #17645

### CODE REVIEW

File: curriculum/challenges/english/02-javascript-algorithms-and-data-structures/regular-expressions/use-capture-groups-to-search-and-replace.english.md [hunk 1]
```diff
@@ -28,7 +28,7 @@ You can also access capture groups in the replacement string with dollar signs (
 
 ## Instructions
 <section id='instructions'>
-Write a regex so that it will search for the string <code>"good"</code>. Then update the <code>replaceText</code> variable to replace <code>"good"</code> with <code>"okey-dokey"</code>.
+Write a regex using capture groups that will search for the words <code>"one"</code>, <code>"two"</code>, and <code>"three"</code>. Then update the <code>replaceText</code> variable to replace <code>"one two three"</code> with <code>"three two one"</code>.
```
Comment (RandellDawson) [inline curriculum/challenges/english/02-javascript-algorithms-and-data-structures/regular-expressions/use-capture-groups-to-search-and-replace.english.md]:
```suggestion
Write a regex using capture groups that will replace the string "one two three" with the string "three two one" and assign the result to the <code>replaceText</code> variable.
```

**Review by RandellDawson — CHANGES_REQUESTED** [review]
@PartyLich In addition to the change I have suggested, there should be an additional test which validates at least three capture groups were used.  Otherwise, a simple replace would solve the challenge without even using regex.

File: curriculum/challenges/english/02-javascript-algorithms-and-data-structures/regular-expressions/use-capture-groups-to-search-and-replace.english.md [hunk 2]
```diff
@@ -38,10 +38,12 @@ Write a regex so that it will search for the string <code>"good"</code>. Then up
...
   - text: You should use <code>.replace()</code> to search and replace.
     testString: assert(code.match(/\.replace\(.*\)/));
-  - text: Your regex should change <code>"This sandwich is good."</code> to <code>"This sandwich is okey-dokey."</code>
-    testString: assert(result == "This sandwich is okey-dokey." && replaceText === "okey-dokey");
+  - text: Your regex should change <code>"one two three"</code> to <code>"three two one"</code>
+    testString: assert(result === "three two one");
   - text: You should not change the last line.
     testString: assert(code.match(/result\s*=\s*huhText\.replace\(.*?\)/));
+  - text: You should use at least three capture groups.
+    testString: assert(code.match(/fixRegex\s*=\s*\/(\(.*?\).*?){3,}\//));
```
Comment (RandellDawson) [inline curriculum/challenges/english/02-javascript-algorithms-and-data-structures/regular-expressions/use-capture-groups-to-search-and-replace.english.md]: The regex for your match would incorrectly pass the one test if someone used the following code.  There are only two capture groups in the following regex.
```js
let fixRegex = /(\w+)\(\w+\)(\w+)/;
```

Courtesy of [Stack Overflow](https://stackoverflow.com/questions/16046620/regex-to-count-the-number-of-capturing-groups-in-a-regex), we can use the following assert:
```suggestion
    testString: assert((new RegExp(fixRegex.source + '|')).exec('').length - 1 >= 3);
```

File: curriculum/challenges/english/02-javascript-algorithms-and-data-structures/regular-expressions/use-capture-groups-to-search-and-replace.english.md [hunk 3]
```diff
@@ -38,10 +38,12 @@ Write a regex so that it will search for the string <code>"good"</code>. Then up
 tests:
   - text: You should use <code>.replace()</code> to search and replace.
     testString: assert(code.match(/\.replace\(.*\)/));
-  - text: Your regex should change <code>"This sandwich is good."</code> to <code>"This sandwich is okey-dokey."</code>
-    testString: assert(result == "This sandwich is okey-dokey." && replaceText === "okey-dokey");
+  - text: Your regex should change <code>"one two three"</code> to <code>"three two one"</code>
+    testString: assert(result === "three two one");
   - text: You should not change the last line.
     testString: assert(code.match(/result\s*=\s*huhText\.replace\(.*?\)/));
+  - text: You should use at least three capture groups.
```
Comment (RandellDawson) [inline curriculum/challenges/english/02-javascript-algorithms-and-data-structures/regular-expressions/use-capture-groups-to-search-and-replace.english.md]:
```suggestion
  - text: <fixRegex> should use at least three capture groups.
```

**Review by RandellDawson — CHANGES_REQUESTED** [review]
See my suggested changes.

File: curriculum/challenges/english/02-javascript-algorithms-and-data-structures/regular-expressions/use-capture-groups-to-search-and-replace.english.md (hunk 2 above)
Comment (PartyLich) [reply curriculum/challenges/english/02-javascript-algorithms-and-data-structures/regular-expressions/use-capture-groups-to-search-and-replace.english.md]: That's an interesting approach.

File: curriculum/challenges/english/02-javascript-algorithms-and-data-structures/regular-expressions/use-capture-groups-to-search-and-replace.english.md [hunk 4]
```diff
@@ -38,10 +38,12 @@ Write a regex so that it will search for the string <code>"good"</code>. Then up
 tests:
   - text: You should use <code>.replace()</code> to search and replace.
     testString: assert(code.match(/\.replace\(.*\)/));
-  - text: Your regex should change <code>"This sandwich is good."</code> to <code>"This sandwich is okey-dokey."</code>
-    testString: assert(result == "This sandwich is okey-dokey." && replaceText === "okey-dokey");
+  - text: Your regex should change <code>"one two three"</code> to <code>"three two one"</code>
+    testString: assert(result === "three two one");
   - text: You should not change the last line.
     testString: assert(code.match(/result\s*=\s*huhText\.replace\(.*?\)/));
```
Comment (RandellDawson) [inline curriculum/challenges/english/02-javascript-algorithms-and-data-structures/regular-expressions/use-capture-groups-to-search-and-replace.english.md]:
```suggestion
    testString: assert(code.match(/result\s*=\s*str\.replace\(.*?\)/));
```

File: curriculum/challenges/english/02-javascript-algorithms-and-data-structures/regular-expressions/use-capture-groups-to-search-and-replace.english.md [hunk 5]
```diff
@@ -53,7 +55,7 @@ tests:
 <div id='js-seed'>
 
 ```js
-let huhText = "This sandwich is good.";
+let huhText = "one two three";
```
Comment (RandellDawson) [inline curriculum/challenges/english/02-javascript-algorithms-and-data-structures/regular-expressions/use-capture-groups-to-search-and-replace.english.md]:
```suggestion
let str = "one two three";
```

File: curriculum/challenges/english/02-javascript-algorithms-and-data-structures/regular-expressions/use-capture-groups-to-search-and-replace.english.md [hunk 6]
```diff
@@ -69,9 +71,9 @@ let result = huhText.replace(fixRegex, replaceText);
 <section id='solution'>
 
 ```js
-let huhText = "This sandwich is good.";
-let fixRegex = /good/g; // Change this line
-let replaceText = "okey-dokey"; // Change this line
+let huhText = "one two three";
```
Comment (RandellDawson) [inline curriculum/challenges/english/02-javascript-algorithms-and-data-structures/regular-expressions/use-capture-groups-to-search-and-replace.english.md]:
```suggestion
let str = "one two three";
```

**Review by RandellDawson — CHANGES_REQUESTED** [review]
@PartyLich Let's change the `huhText` which means nothing to `str`.  

Also, you need to figure out a way to not allow a solution like the following which does not use replacement patterns:
```js
let str = "one two three";
let fixRegex = /(\w+) (\w+) (\w+)/; // Change this line
let replaceText = "three two one"; // Change this line
let result = str.replace(fixRegex, replaceText);
```

**Comment by PartyLich** [comment]
We could use a regex to ensure there is some instance of `$n` with n in the 1-99 range, but I'm not certain at the moment what new holes that approach would have. Do we care that they use >1 submatch in the final replacement? It would make sense to me to use all 3 of the mandated capture groups in the replacement as well, but idk if we want to enforce that.

**Comment by PartyLich** [comment]
new test added.
*passing:*
```js
let str = "one two three";
let fixRegex = /(\w+) (\w+) (\w+)/; // Change this line
let replaceText = "$3 $2 $1"; // Change this line
let result = str.replace(fixRegex, replaceText);
```
```js
…
let replaceText = "three $2 $1"; // Change this line
…
```
```js
…
let replaceText = "three two $1"; // Change this line
…
```

*failing*
```js
let str = "one two three";
let fixRegex = /(\w+) (\w+) (\w+)/; // Change this line
let replaceText = "three two one"; // Change this line
let result = str.replace(fixRegex, replaceText);
```

"Parenthesized submatch string" is interesting wording...but it matches MDN should a student happen to use that source for research/help. I'm obviously open to suggested replacements.

**Comment by RandellDawson** [comment]
I do not like the fact that the following pass:
```js
let replaceText = "three $2 $1"; // Change this line
```

```js
let replaceText = "three two $1"; // Change this line
```
The user should only be using spaces and $n for the replacement and not hard-coded values.

**Comment by PartyLich** [comment]
> I do not like the fact that the following pass:
> 
> ```js
> let replaceText = "three $2 $1"; // Change this line
> ```
> 
> ```js
> let replaceText = "three two $1"; // Change this line
> ```
> 
> The user should only be using spaces and $n for the replacement and not hard-coded values.

That's why my previous comment included the question "Do we care that they use >1 submatch in the final replacement?". Whether they use 1 or 3 or n, they're still demonstrating that they are familiar with the concept and can utilize it to update a string.

**Comment by RandellDawson** [comment]
Let's make sure they use all 3.

**Review by RandellDawson — APPROVED** [review]

**Comment by RandellDawson** [comment]
OK, I think this one looks good now.

File: curriculum/challenges/english/02-javascript-algorithms-and-data-structures/regular-expressions/use-capture-groups-to-search-and-replace.english.md [hunk 7]
```diff
@@ -28,7 +28,7 @@ You can also access capture groups in the replacement string with dollar signs (
 
 ## Instructions
 <section id='instructions'>
-Write a regex so that it will search for the string <code>"good"</code>. Then update the <code>replaceText</code> variable to replace <code>"good"</code> with <code>"okey-dokey"</code>.
+Write a regex using three capture groups that will replace the string "one two three" with the string "three two one" and assign the result to the <code>replaceText</code> variable.
```
Comment (Manish-Giri) [inline curriculum/challenges/english/02-javascript-algorithms-and-data-structures/regular-expressions/use-capture-groups-to-search-and-replace.english.md]: > `the result to the <code>replaceText</code> variable.` 

Shouldn't this be the `result` variable?

File: curriculum/challenges/english/02-javascript-algorithms-and-data-structures/regular-expressions/use-capture-groups-to-search-and-replace.english.md [hunk 8]
```diff
@@ -38,10 +38,17 @@ Write a regex so that it will search for the string <code>"good"</code>. Then up
...
-  - text: Your regex should change <code>"This sandwich is good."</code> to <code>"This sandwich is okey-dokey."</code>
-    testString: assert(result == "This sandwich is okey-dokey." && replaceText === "okey-dokey");
+  - text: Your regex should change <code>"one two three"</code> to <code>"three two one"</code>
+    testString: assert(result === "three two one");
   - text: You should not change the last line.
-    testString: assert(code.match(/result\s*=\s*huhText\.replace\(.*?\)/));
+    testString: assert(code.match(/result\s*=\s*str\.replace\(.*?\)/));
+  - text: <code>fixRegex</code> should use at least three capture groups.
+    testString: assert((new RegExp(fixRegex.source + '|')).exec('').length - 1 >= 3);
+  - text: <code>replaceText</code> should use parenthesized submatch string(s).
```
Comment (Manish-Giri) [inline curriculum/challenges/english/02-javascript-algorithms-and-data-structures/regular-expressions/use-capture-groups-to-search-and-replace.english.md]: > <code>replaceText</code> should use parenthesized submatch string(s).

If I'm new to regex, going solely by the challenge description, I have no idea what "parenthesized submatch string(s)" means. Probably should be rephrased.

**Review by Manish-Giri — CHANGES_REQUESTED** [review]
In addition to the suggested changes, the challenge instruction should also include what to do with the `fixRegex` and `replaceText` variables in the editor. Currently, the instruction just states - 

> Write a regex using three capture groups that will replace the string "one two three" with the string "three two one" and assign the result to the replaceText variable.

You shouldn't be required to look at the tests to figure out what the variables are for -
> fixRegex should use at least three capture groups.
> replaceText should use parenthesized submatch string(s).

We could use something on the lines of what we had in the previous version - 

> Write a regex so that it will search for the string "good". Then update the replaceText variable to replace "good" with "okey-dokey"

File: curriculum/challenges/english/02-javascript-algorithms-and-data-structures/regular-expressions/use-capture-groups-to-search-and-replace.english.md (hunk 8 above)
Comment (PartyLich) [reply curriculum/challenges/english/02-javascript-algorithms-and-data-structures/regular-expressions/use-capture-groups-to-search-and-replace.english.md]: Any suggestions on what to rephrase it to?  I didn't come across or previously know an alternate term. If I were a new(er) student and making use of the "Read-Search-Ask" slogan that was so prominent on FCC, a google search of the phrase would produce accurate results (for multiple programming languages). If I were me and had the habit of looking up js docs on MDN, the String.prototype.replace() page has this exact wording.

I have no strong feelings about changing it to something else, I just have no idea what would be _better_. I could use some help there.

File: curriculum/challenges/english/02-javascript-algorithms-and-data-structures/regular-expressions/use-capture-groups-to-search-and-replace.english.md (hunk 7 above)
Comment (PartyLich) [reply curriculum/challenges/english/02-javascript-algorithms-and-data-structures/regular-expressions/use-capture-groups-to-search-and-replace.english.md]: Good catch. Fix will be up momentarily.

File: curriculum/challenges/english/02-javascript-algorithms-and-data-structures/regular-expressions/use-capture-groups-to-search-and-replace.english.md (hunk 8 above)
Comment (PartyLich) [reply curriculum/challenges/english/02-javascript-algorithms-and-data-structures/regular-expressions/use-capture-groups-to-search-and-replace.english.md]: An/Or I suppose we could add the preferred terminology to the description. At present it says something to the effect of "use dollar signs", but offers no terminology to concisely describe that.

File: curriculum/challenges/english/02-javascript-algorithms-and-data-structures/regular-expressions/use-capture-groups-to-search-and-replace.english.md [hunk 9]
```diff
@@ -28,7 +28,7 @@ You can also access capture groups in the replacement string with dollar signs (
 
 ## Instructions
 <section id='instructions'>
-Write a regex so that it will search for the string <code>"good"</code>. Then update the <code>replaceText</code> variable to replace <code>"good"</code> with <code>"okey-dokey"</code>.
+Write a regex <code>fixRegex</code> using three capture groups that will search for each word in the string "one two three". Then update the <code>replaceText</code> variable to replace "one two three" with the string "three two one" and assign the result to the <code>result</code> variable. Make sure you utilizing your capture groups in the replacement string using the dollar sign (<code>$</code>) syntax.
```
Comment (Manish-Giri) [inline curriculum/challenges/english/02-javascript-algorithms-and-data-structures/regular-expressions/use-capture-groups-to-search-and-replace.english.md]:
```suggestion
Write a regex <code>fixRegex</code> using three capture groups that will search for each word in the string "one two three". Then update the <code>replaceText</code> variable to replace "one two three" with the string "three two one" and assign the result to the <code>result</code> variable. Make sure you are utilizing capture groups in the replacement string using the dollar sign (<code>$</code>) syntax.
```

File: curriculum/challenges/english/02-javascript-algorithms-and-data-structures/regular-expressions/use-capture-groups-to-search-and-replace.english.md (hunk 8 above)
Comment (Manish-Giri) [reply curriculum/challenges/english/02-javascript-algorithms-and-data-structures/regular-expressions/use-capture-groups-to-search-and-replace.english.md]: Since you have now edited the instructions to add a little detail about the `$` syntax, perhaps this could be (re)used to simplify the terminology - 

> text: <code>replaceText</code> should use parenthesized submatch string(s) (ie. the nth parenthesized submatch string, `$n`, corresponds to the nth capture group).

The MDN docs on `.replace()` uses this terminology.

**Review by Manish-Giri — CHANGES_REQUESTED** [review]
Minor grammar fix.

**Review by Manish-Giri — APPROVED** [review]
LGTM. Thank You for your contribution.
